internal/redirecttraffic/redirect_traffic.go

```diff
@@ -0,0 +1,205 @@
+	BindPort           int    `mapstructure:"bind_port"`
+	PrometheusBindAddr string `mapstructure:"envoy_prometheus_bind_addr"`
+	StatsBindAddr      string `mapstructure:"envoy_stats_bind_addr"`
+}
+
+type TrafficRedirectionCfg struct {
+	ProxySvc *api.AgentService
+
+	ConsulServerAddress  string
+	ClusterARN           string
+	ProxyHealthCheckPort int
+
+	EnableConsulDNS      bool
+	ExcludeInboundPorts  []int
+	ExcludeOutboundPorts []int
+	ExcludeOutboundCIDRs []string
+	ExcludeUIDs          []string
+
+	iptablesCfg iptables.Config
+
+	// Fields used only for unit tests
+	iptablesProvider iptables.Provider
+}
+
+type TrafficRedirectionProvider interface {
+	Apply() error
+
+	// Used to fetch the resultant config in unit tests
+	config() iptables.Config
+}
+
+type TrafficRedirectionOpts func(*TrafficRedirectionCfg)
+
+func WithIPTablesProvider(provider iptables.Provider) TrafficRedirectionOpts {
+	return func(c *TrafficRedirectionCfg) {
+		c.iptablesProvider = provider
+	}
+}
+
+func New(cfg *config.Config, proxySvc *api.AgentService, consulServerAddress, clusterARN string, proxyHealthCheckPort int, opts ...TrafficRedirectionOpts) TrafficRedirectionProvider {
+	input := &TrafficRedirectionCfg{
+		ProxySvc:             proxySvc,
+		ConsulServerAddress:  consulServerAddress,
+		ClusterARN:           clusterARN,
+		ProxyHealthCheckPort: proxyHealthCheckPort,
+		EnableConsulDNS:      cfg.ConsulDNSEnabled(),
+		ExcludeInboundPorts:  cfg.TransparentProxy.ExcludeInboundPorts,
+		ExcludeOutboundPorts: cfg.TransparentProxy.ExcludeOutboundPorts,
+		ExcludeOutboundCIDRs: cfg.TransparentProxy.ExcludeOutboundCIDRs,
+		ExcludeUIDs:          cfg.TransparentProxy.ExcludeUIDs,
+	}
+
+	for _, opt := range opts {
+		opt(input)
+	}
+
+	return input
+}
+
+// applyTrafficRedirectionRules creates and applies traffic redirection rules with
+// the help of iptables
+//
+// iptables.Config:
+//
+//	ConsulDNSIP: Consul Dataplane's DNS server (i.e. localhost)
+//	ConsulDNSPort: Consul Dataplane's DNS server's bind port
+//	ProxyUserID: a constant set by default in the mesh-task module for the Consul dataplane's container
+//	ProxyInboundPort: the default inbound port or bind port
+//	ProxyOutboundPort: default transparent proxy outbound port
+//	ExcludeInboundPorts: prometheus, envoy stats, expose paths and `transparentProxy.excludeInboundPorts`
+//	ExcludeOutboundPorts: `transparentProxy.excludeOutboundPorts` in CONSUL_ECS_CONFIG_JSON
+//	ExcludeOutboundCIDRs: TaskMeta IP, Consul Server IP and `transparentProxy.excludeOutboundCIDRs` in CONSUL_ECS_CONFIG_JSON
+//	ExcludeUIDs: `transparentProxy.excludeUIDs` in CONSUL_ECS_CONFIG_JSON
+func (c *TrafficRedirectionCfg) Apply() error {
+	if c.ProxySvc == nil {
+		return fmt.Errorf("proxy service details are required to enable traffic redirection")
+	}
+
+	// Decode proxy's opaque config
+	var trCfg trafficRedirectProxyConfig
+	if err := mapstructure.WeakDecode(c.ProxySvc.Proxy.Config, &trCfg); err != nil {
```

Comment:
It can't be `nil` for a proxy service. We could may be add this check when passing the proxySvc to the `Apply` function in the caller. Those changes will be added in a followup PR.